### PR TITLE
[FIX] web: use back button don't break filters in search view

### PR DESCRIPTION
**Pull request by jpp-odoo**
[FIX] web: use back button don't break filters in search view

Open a kanban view or a list view;
Enter a filter in the search view;
Open a record;
Go back to the multi-record view using the browser (back button).

Before this commit, the previously entered filter is not there anymore.
Now, the filter stays. To achieve this, before leaving the multi-record view, the global state is saved on the history state.

To be able to push a key (the global state) on the state to be saved on the history and not shown in the browser, a new way to add a key/value to the state,
but omit it on the URL was developed. Also, a way to push the state synchronous was added.

### CODE REVIEW

File: addons/web/static/tests/core/router.test.js [hunk 1]
```diff
@@ -1361,7 +1361,7 @@ describe("pushState", () => {
     test("can lock keys", async () => {
         createRouter();
 
-        router.addLockedKey(["k1"]);
+        router.addLockedKey("k1");
```
Comment (aab-odoo) [inline addons/web/static/tests/core/router.test.js:1376]: ["k1"] works?

File: addons/web/static/src/core/browser/router.js [hunk 2]
```diff
@@ -366,6 +368,7 @@ export const router = {
     replaceState: makeDebouncedPush("replace"),
     cancelPushes: () => browser.clearTimeout(pushTimeout),
     addLockedKey: (key) => _lockedKeys.add(key),
+    addKeyToOmit: (key) => _pathKeysToOmit.add(key),
```
Comment (aab-odoo) [inline addons/web/static/src/core/browser/router.js:378]: this commit is quite elegant, I simply don't like the name of this function.

File: addons/web/static/tests/core/router.test.js [hunk 3]
```diff
@@ -1358,6 +1358,18 @@ describe("pushState", () => {
...
+        createRouter();
+
+        expect(router.current).toEqual({});
+
+        router.pushState({ k1: 2 }, { sync: true });
+        expect(router.current).toEqual({ k1: 2 });
+
+        router.pushState({ k1: 3 }, { sync: true });
+        expect(router.current).toEqual({ k1: 3 });
+    });
```
Comment (aab-odoo) [inline addons/web/static/tests/core/router.test.js:1371]: router.pushState({ k1: 4 });
router.pushState({ k2: 1 });
expect(router.current).toEqual(???);

File: addons/web/static/src/webclient/actions/action_service.js [hunk 4]
```diff
@@ -961,11 +969,16 @@ export function makeActionManager(env, router = _router) {
...
         if (currentController && currentController.getGlobalState) {
-            currentController.action.globalState = Object.assign(
+            const globalState = Object.assign(
                 {},
                 currentController.action.globalState,
                 currentController.getGlobalState() // what if this = {}?
             );
+
+            currentController.action.globalState = globalState;
+            if (Object.keys(globalState).length) {
```
Comment (aab-odoo) [inline addons/web/static/src/webclient/actions/action_service.js:979]: is this if necessary?

File: addons/web/static/tests/legacy/webclient/actions/load_state_tests.js [hunk 5]
```diff
@@ -45,7 +45,11 @@ const logHistoryInteractions = (assert) => {
             return super.pushState(state, _, url);
         },
         replaceState(state, _, url) {
-            assert.step(`replaceState ${url}`);
+            if (browser.location.href === url) {
+                assert.step(`Update the state without updating URL`);
```
Comment (aab-odoo) [inline addons/web/static/tests/legacy/webclient/actions/load_state_tests.js:49]: maybe step the state's keys (not values) ?

File: addons/web/static/src/core/browser/router.js (hunk 2 above)
Comment (aab-odoo) [reply addons/web/static/src/core/browser/router.js:378]: suggestion: hideKeyInUrl

File: addons/web/static/src/core/browser/router.js [hunk 6]
```diff
@@ -228,6 +228,7 @@ let state;
 let pushTimeout;
 let pushArgs;
 let _lockedKeys;
+let _hideKeysFromUrl = new Set();
```
Comment (aab-odoo) [inline addons/web/static/src/core/browser/router.js:231]: _hiddenKeysFromUrl

**Review by aab-odoo — APPROVED** [review]
robodoo delegate+

**Comment by aab-odoo** [comment]
robodoo rebase-merge

**Comment by jpp-odoo** [comment]
@robodoo r+

**Pull request closed without merging**